Feat/permissioned deployments

RomanHiden opened the pull request:
## What ❔
This PR is adding new composite WhitelistedDeployPoolSink layer 

## Why ❔

This  allows to make chain permission ed. allowing only whitelisted list of addresses deploy contract

## Is this a breaking change?
- [ ] Yes
- [X ] No

## Operational changes
TODO

## Checklist

<!-- Check your PR fulfills the following items. -->
<!-- For draft PRs check the boxes as you complete them. -->

- [ Y] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ N] Tests for the changes have been added / updated.
- [ N] Documentation comments have been added / updated.
- [ Y] Code has been formatted via `zkstack dev fmt` and `zkstack dev lint`.

RomanHiden committed:
added dal; query database

RomanHiden committed:
added dal; query database

RomanHiden committed:
original package.json

RomanHiden committed:
added VmExecutionMetrics is_contrac_deployed

RomanHiden committed:
merged with upstream

RomanHiden committed:
formated

RomanHiden committed:
implemented tx_sink configuration

RomanHiden committed:
implemented tx_sink configuration

RomanHiden committed:
unneded file

RomanHiden committed:
improved dal speed; self review

RomanHiden committed:
new sql query

RomanHiden committed:
remove comments

RomanHiden committed:
removed accidental modifications

RomanHiden committed:
leaving comment

RomanHiden committed:
leaving package.json as is

dimazhornyk commented on core/lib/env_config/src/tx_sink.rs:
```diff
@@ -0,0 +1,35 @@
+    fn from_env() {
+        let mut lock = MUTEX.lock();
+        let config = r#"
+            use_whitelisted_sink=false
```
this seems like it might not work, you probably need to use `TX_SINK_USE_WHITELISTED_SINK` there
and please change it to true, so you are testing a non-default condition

dimazhornyk commented on core/lib/config/src/configs/tx_sink.rs:
```diff
@@ -0,0 +1,6 @@
+
+#[derive(Debug, Deserialize, Clone, PartialEq)]
+pub struct TxSinkConfig {
+    pub use_whitelisted_sink: Option<bool>,
```
nit: I'd rather call the feature an `deployment_allowlist_sink`, so here it could be `use_deployment_allowlist_sink`

dimazhornyk replied:
`Option` can be replaced here with `#[serde(default...`, this way we kind of define the default behavior globally

dimazhornyk commented on core/lib/vm_interface/src/types/outputs/statistic.rs:
```diff
@@ -251,6 +253,7 @@ impl ops::Add for VmExecutionMetrics {
             computational_gas_used: self.computational_gas_used + other.computational_gas_used,
             pubdata_published: self.pubdata_published + other.pubdata_published,
             circuit_statistic: self.circuit_statistic + other.circuit_statistic,
+            is_contract_deployed: self.is_contract_deployed || other.is_contract_deployed,
```
nit: this doesn't look right, maybe we can track a deployment count instead, and then we can implement a proper `add`

dimazhornyk commented on core/node/api_server/src/tx_sender/result.rs:
```diff
@@ -69,6 +69,8 @@ pub enum SubmitTxError {
     #[error("transaction failed block.timestamp assertion")]
     FailedBlockTimestampAssertion,
+    #[error("sender address {0} is in allow list")]
+    SenderInAllowList(Address),
```
if it is an error - then it should be `SenderNotInAllowList`, right?

dimazhornyk commented on core/node/api_server/src/tx_sender/whitelisted_deploy_pool_sink.rs:
```diff
@@ -0,0 +1,77 @@
+/// Wrapper for the master DB pool that allows submitting transactions to the mempool.
+#[derive(Debug)]
+pub struct WhitelistedDeployPoolSink {
+    master_pool_sync: MasterPoolSink,
```
nit: master_pool_sync -> master_pool_sink

dimazhornyk commented on core/node/api_server/src/tx_sender/whitelisted_deploy_pool_sink.rs:
```diff
@@ -0,0 +1,77 @@
+            // Only enforce the allow-list check if the deployer contract was called
+            // and the VM actually deployed a contract.
+            if contract_address == CONTRACT_DEPLOYER_ADDRESS
+                && execution_metrics.vm.is_contract_deployed
```
if the proxy/factory is called - then `contract_address` won't be equal to `CONTRACT_DEPLOYER_ADDRESS`, but the deployment will still happen

I'd say `execution_metrics.vm.is_contract_deployed` is enough here

dimazhornyk commented on core/node/api_server/src/tx_sender/whitelisted_deploy_pool_sink.rs:
```diff
@@ -0,0 +1,77 @@
+                    .unwrap();
+
+                if is_allowed {
+                    tracing::info!(
```
let's use `tracing::debug` here to not overload the info logs

dimazhornyk commented on core/node/api_server/src/tx_sender/whitelisted_deploy_pool_sink.rs:
```diff
@@ -0,0 +1,77 @@
+                    .contracts_deploy_allow_list_dal()
+                    .is_address_allowed(&initiator)
+                    .await
+                    .unwrap();
```
I'm not sure we have to panic here in case of error, would it be possible to replace `unwrap()` with `.map_err(...)?`?

RomanHiden committed:
Merge branch 'main' into feat/permissioned-deployments

StanislavBreadless committed:
feat(v27): Use latest branch of release-v27 (#3713)

## What ❔

<!-- What are the changes this PR brings about? -->
<!-- Example: This PR adds a PR template to the repo. -->
<!-- (For bigger PRs adding more context is appreciated) -->

## Why ❔

<!-- Why are these changes done? What goal do they contribute to? What
are the principles behind them? -->
<!-- The `Why` has to be clear to non-Matter Labs entities running their
own ZK Chain -->
<!-- Example: PR templates ensure PR reviewers, observers, and future
iterators are in context about the evolution of repos. -->

## Is this a breaking change?
- [ ] Yes
- [ ] No

## Operational changes
<!-- Any config changes? Any new flags? Any changes to any scripts? -->
<!-- Please add anything that non-Matter Labs entities running their own
ZK Chain may need to know -->

## Checklist

<!-- Check your PR fulfills the following items. -->
<!-- For draft PRs check the boxes as you complete them. -->

- [ ] PR title corresponds to the body of PR (we generate changelog
entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zkstack dev fmt` and `zkstack dev
lint`.

haraldh committed:
fix(zkstack): add `--locked` to `cargo sqlx prepare` (#3300)

## What ❔

Add `--locked` to `cargo sqlx prepare`.

## Why ❔

This prevents accidental `Cargo.lock` updates.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog
entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [x] Code has been formatted via `zkstack dev fmt` and `zkstack dev
lint`.

Signed-off-by: Harald Hoyer <harald@matterlabs.dev>

RomanHiden committed:
review fixes

RomanHiden committed:
small improvements; default config values

slowli committed:
fix(vm): Fix another VM divergence in validation (#3675)

## What ❔

Fixes VM divergence during account validation caused by the fast VM
validation tracer not taking adjacent storage slots into account.

## Why ❔

We don't want to have VM divergences.

## Is this a breaking change?

- [ ] Yes
- [x] No

## Operational changes

No operational changes.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog
entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zkstack dev fmt` and `zkstack dev
lint`.

EmilLuta committed:
docs: Adds docs on linting contracts (#3731)

Small doc contribution for local development. Helps with running lints
on contracts locally.

RomanHiden committed:
removed leftover comment

RomanHiden committed:
review changes

RomanHiden committed:
changed to main

RomanHiden committed:
Merge branch 'main' into feat/permissioned-deployments
